kernel/src/utilities/streaming_process_slice.rs

```diff
@@ -0,0 +1,269 @@
+    ///   flags field and the offset field (5 bytes).
+    pub fn get_meta(&self) -> Result<StreamingProcessSliceMeta, ErrorCode> {
+        let mut flags_bytes = [0_u8];
+        self.slice
+            .get(Self::RANGE_FLAGS)
+            .ok_or(ErrorCode::SIZE)?
+            .copy_to_slice_or_err(&mut flags_bytes)
+            .map_err(|_| ErrorCode::SIZE)?;
+
+        let flags: LocalRegisterCopy<u8, StreamingProcessSliceFlags::Register> =
+            LocalRegisterCopy::new(flags_bytes[0]);
+
+        let version = match flags.read_as_enum(StreamingProcessSliceFlags::VERSION) {
+            Some(v @ StreamingProcessSliceFlags::VERSION::Value::V0) => v,
+            _ => return Err(ErrorCode::INVAL),
+        };
+
+        if flags.read_as_enum(StreamingProcessSliceFlags::RESERVED)
+            != Some(StreamingProcessSliceFlags::RESERVED::Value::RESERVED0)
+        {
+            return Err(ErrorCode::INVAL);
+        }
+
+        let mut offset_bytes = [0_u8; Self::RANGE_OFFSET.end - Self::RANGE_OFFSET.start];
+        self.slice
+            .get(Self::RANGE_OFFSET)
+            .ok_or(ErrorCode::SIZE)?
+            .copy_to_slice_or_err(&mut offset_bytes)
+            .map_err(|_| ErrorCode::SIZE)?;
+
+        Ok(StreamingProcessSliceMeta {
+            version,
+            halt: flags.read(StreamingProcessSliceFlags::HALT) != 0,
+            exceeded: flags.read(StreamingProcessSliceFlags::EXCEEDED) != 0,
+            offset: u32::from_ne_bytes(offset_bytes),
+        })
+    }
+
+    /// Write updated metadata (`flags` and `offset`) back to the underlying
+    /// slice.
+    ///
+    /// This function does not perform any sanity checks on the `meta` argument.
+    /// In particular, users of this function must ensure that they previously
+    /// extracted the written-back [`StreamingProcessSliceMeta`] argument from
+    /// the buffer, do not modify the version, do not change any flags that are
+    /// controlled by the process or otherwise violate the protocol, and
+    /// correctly increment the `offset` value.
+    ///
+    /// - `SIZE`: if the underlying slice is not large enough to fit the
+    ///   flags field and the offset field (5 bytes).
+    pub fn set_meta(&self, meta: StreamingProcessSliceMeta) -> Result<(), ErrorCode> {
+        // Write the offset first, to avoid modifying the buffer if it's too
+        // small to fit the offset, but large enough to hold the flags byte:
+        let offset_bytes = u32::to_ne_bytes(meta.offset);
+        self.slice
+            .get(Self::RANGE_OFFSET)
+            .ok_or(ErrorCode::SIZE)?
+            .copy_from_slice_or_err(&offset_bytes)
+            .map_err(|_| ErrorCode::SIZE)?;
+
+        let flags_bytes: [u8; 1] = [(StreamingProcessSliceFlags::VERSION.val(meta.version as u8)
+            + StreamingProcessSliceFlags::RESERVED::RESERVED0
+            + StreamingProcessSliceFlags::HALT.val(meta.halt as u8)
```

Comment:
Do we want to allow the kernel to write the `HALT` flag?

Reply:
No, it generally shouldn't. Instead this method is simply used to write back a modified header, and we should promise to never change the `flag` value in `meta.flag` from when it was read from the buffer, and never retain `StreamingProcessSliceMeta` values beyond switching back to a process.

We'll ensure this by making these methods private, compare https://github.com/tock/tock/pull/4208/files#r1811268599